sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Processor/PartitionLoadBalancer.cs

```diff
@@ -0,0 +1,443 @@
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Messaging.EventHubs.Core;
+using Azure.Messaging.EventHubs.Errors;
+
+namespace Azure.Messaging.EventHubs.Processor
+{
+    internal class PartitionLoadBalancer
+    {
+        /// <summary>The random number generator to use for a specific thread.</summary>
+        private static readonly ThreadLocal<Random> RandomNumberGenerator = new ThreadLocal<Random>(() => new Random(Interlocked.Increment(ref s_randomSeed)), false);
+
+        /// <summary>The seed to use for initializing random number generated for a given thread-specific instance.</summary>
+        private static int s_randomSeed = Environment.TickCount;
+
+        /// <summary>
+        ///   Interacts with the storage system with responsibility for creation of checkpoints and for ownership claim.
+        /// </summary>
+        ///
+        private readonly PartitionManager StorageManager;
+
+        /// <summary>
+        ///   The set of partition ownership this event processor owns.  Partition ids are used as keys.
+        /// </summary>
+        ///
+        private Dictionary<string, PartitionOwnership> InstanceOwnership { get; set; } = new Dictionary<string, PartitionOwnership>();
+
+        /// <summary>
+        ///   A partition distribution dictionary, mapping an owner's identifier to the amount of partitions it owns and its list of partitions.
+        /// </summary>
+        ///
+        private readonly Dictionary<string, List<PartitionOwnership>> ActiveOwnershipWithDistribution = new Dictionary<string, List<PartitionOwnership>>();
+
+        /// <summary>
+        ///   The fully qualified Event Hubs namespace that the processor is associated with.  This is likely
+        ///   to be similar to <c>{yournamespace}.servicebus.windows.net</c>.
+        /// </summary>
+        ///
+        public readonly string FullyQualifiedNamespace;
+
+        /// <summary>
+        ///   The name of the Event Hub that the processor is connected to, specific to the
+        ///   Event Hubs namespace that contains it.
+        /// </summary>
+        ///
+        public readonly string EventHubName;
+
+        /// <summary>
+        ///   The name of the consumer group this event processor is associated with.  Events will be
+        ///   read only in the context of this group.
+        /// </summary>
+        ///
+        public readonly string ConsumerGroup;
+
+        /// <summary>
+        ///   A unique name used to identify this event processor.
+        /// </summary>
+        ///
+        public readonly string Identifier;
+
+        /// <summary>
+        ///   The minimum amount of time to be elapsed between two load balancing verifications.
+        /// </summary>
+        ///
+        public virtual TimeSpan LoadBalanceUpdate => TimeSpan.FromSeconds(10);
```

Comment:
nit:  The name here is a bit awkward (and has been for a bit.)    Maybe consider something like `UpdateInterval`  or `UpdateLoadBalancingAfter` or `RebalanceAfter` or something else?

Reply:
I went with `LoadBalanceInterval`